File: aether_messenger/android/app/src/main/java/com/aetherq/messenger/crypto/ContactCard.kt
package com.aetherq.messenger.crypto

import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/** QR-kod orqali almashiladigan ochiq kalitlar to'plami (Signal safety-number uslubida). */
@Serializable
data class ContactCard(
    val userId: String,
    val edPk: String,
    val xPk: String,
    val mlkemPk: String,
)

object ContactCardCodec {
    private val json = Json { ignoreUnknownKeys = true }

    fun encode(card: ContactCard): String = json.encodeToString(card)

    fun decode(raw: String): ContactCard = json.decodeFromString(raw)

    fun fromIdentity(identity: Identity): ContactCard =
        ContactCard(
            userId = identity.userId,
            edPk = identity.edPk.toHex(),
            xPk = identity.xPk.toHex(),
            mlkemPk = identity.mlkemPk.toHex(),
        )
}

File: aether_messenger/windows/AetherQMessenger/Crypto/ContactCard.cs
using System;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace AetherQMessenger.Crypto;

/// <summary>
/// Kontakt ochiq kalitlari — mobil (Android) versiyadagi <c>ContactCard</c> bilan bir xil
/// JSON shakli (<c>userId</c>/<c>edPk</c>/<c>xPk</c>/<c>mlkemPk</c>, kichik harfli hex) —
/// shu tufayli QR-kod matni ikkala klient o'rtasida almashtirilishi mumkin.
/// </summary>
public sealed class ContactCard
{
    [JsonPropertyName("userId")]
    public required string UserId { get; init; }

    [JsonPropertyName("edPk")]
    public required string EdPk { get; init; }

    [JsonPropertyName("xPk")]
    public required string XPk { get; init; }

    [JsonPropertyName("mlkemPk")]
    public required string MlkemPk { get; init; }
}

public static class ContactCardCodec
{
    public static string Encode(ContactCard card) => JsonSerializer.Serialize(card);

    public static ContactCard Decode(string raw) =>
        JsonSerializer.Deserialize<ContactCard>(raw.Trim())
        ?? throw new FormatException("Yaroqsiz kontakt-karta matni");

    public static ContactCard FromIdentity(Identity identity) =>
        new()
        {
            UserId = identity.UserId,
            EdPk = Convert.ToHexString(identity.EdPk).ToLowerInvariant(),
            XPk = Convert.ToHexString(identity.XPk).ToLowerInvariant(),
            MlkemPk = Convert.ToHexString(identity.MlkemPk).ToLowerInvariant(),
        };
}
